File: adapters/macos/Sources/kakao-macos-bridge/KoreanTime.swift
import Foundation

/// KakaoTalk's list and message views expose time as Korean locale labels
/// (`"오전 11:17"`, `"오후 12:12"`, `"어제"`, `"2026년 8월 1일"`), never ISO.
/// This does a best-effort conversion to an ISO 8601 UTC string; when a label
/// carries no usable time it returns `""` (the contract allows an empty `at`
/// for list previews — see `docs/adapter-contract.md`).
enum KoreanTime {
    private static var cal: Calendar {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone.current
        return c
    }

    private static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        f.timeZone = TimeZone(identifier: "UTC")
        return f
    }()

    /// `label` is a KakaoTalk time string. `now` is injectable for tests.
    static func toISO(_ label: String, now: Date = Date()) -> String {
        let s = label.trimmingCharacters(in: .whitespaces)
        guard !s.isEmpty else { return "" }

        // "오전/오후 H:MM"  ->  today (or yesterday if `dayOffset` given elsewhere)
        if let hm = parseHourMinute(s) {
            return iso.string(from: combine(day: now, hour: hm.0, minute: hm.1))
        }
        // "어제"
        if s == "어제" {
            let y = cal.date(byAdding: .day, value: -1, to: now) ?? now
            return iso.string(from: cal.startOfDay(for: y))
        }
        // "YYYY년 M월 D일"
        if let d = parseYMD(s) {
            return iso.string(from: d)
        }
        return ""
    }

    /// Parse "오전 11:17" / "오후 12:12" -> 24h (hour, minute).
    static func parseHourMinute(_ s: String) -> (Int, Int)? {
        let ampm: Int
        var rest = s
        if s.hasPrefix("오전") { ampm = 0; rest = String(s.dropFirst(2)) }
        else if s.hasPrefix("오후") { ampm = 12; rest = String(s.dropFirst(2)) }
        else { return nil }

        let parts = rest.trimmingCharacters(in: .whitespaces).split(separator: ":")
        guard parts.count == 2, var h = Int(parts[0]), let m = Int(parts[1]) else { return nil }
        // 오전 12:xx -> 00:xx ; 오후 12:xx -> 12:xx
        if h == 12 { h = 0 }
        return (h + ampm, m)
    }

    private static func parseYMD(_ s: String) -> Date? {
        let digits = s.split(whereSeparator: { !$0.isNumber }).compactMap { Int($0) }
        guard digits.count == 3 else { return nil }
        var dc = DateComponents()
        dc.year = digits[0]; dc.month = digits[1]; dc.day = digits[2]
        return cal.date(from: dc)
    }

    private static func combine(day: Date, hour: Int, minute: Int) -> Date {
        var dc = cal.dateComponents([.year, .month, .day], from: day)
        dc.hour = hour; dc.minute = minute; dc.second = 0
        return cal.date(from: dc) ?? day
    }
}

File: adapters/windows/src/korean_time.rs
//! KakaoTalk's list and message views expose time as Korean locale labels
//! (`"오전 11:17"`, `"오후 12:12"`, `"어제"`), never ISO. Best-effort conversion
//! to an ISO 8601 UTC string; `""` when a label carries no usable time.
//!
//! Mirrors the macOS adapter's `KoreanTime` — the two adapters must produce
//! identical `at` values for identical labels.
//!
//! NOTE: KakaoTalk Windows locale strings need verification against a real
//! build via `--dump-tree`.

use time::format_description::well_known::Rfc3339;
use time::{Duration, OffsetDateTime, Time, UtcOffset};

fn local_offset() -> UtcOffset {
    UtcOffset::current_local_offset().unwrap_or(UtcOffset::UTC)
}

fn iso_utc(dt: OffsetDateTime) -> String {
    dt.to_offset(UtcOffset::UTC)
        .replace_nanosecond(0)
        .unwrap_or(dt)
        .format(&Rfc3339)
        .unwrap_or_default()
}

/// Parse "오전 11:17" / "오후 12:12" (or "1\n오전 1:01") -> 24h (hour, minute).
pub fn parse_hour_minute(raw: &str) -> Option<(u8, u8)> {
    let line = raw.lines().last().unwrap_or(raw).trim();
    let (ampm, rest) = if let Some(r) = line.strip_prefix("오전") {
        (0u8, r)
    } else {
        let r = line.strip_prefix("오후")?;
        (12u8, r)
    };
    let (h, m) = rest.trim().split_once(':')?;
    let mut h: u8 = h.trim().parse().ok()?;
    let m: u8 = m.trim().parse().ok()?;
    if h == 12 {
        h = 0;
    }
    if h + ampm > 23 || m > 59 {
        return None;
    }
    Some((h + ampm, m))
}

/// `label` -> ISO 8601 UTC string, or `""`. `now` is injectable for tests.
pub fn to_iso(label: &str, now: OffsetDateTime) -> String {
    let line = label.lines().last().unwrap_or(label).trim();
    if line.is_empty() {
        return String::new();
    }
    let local_now = now.to_offset(local_offset());

    if let Some((h, m)) = parse_hour_minute(line) {
        let local = local_now.replace_time(Time::from_hms(h, m, 0).unwrap_or(Time::MIDNIGHT));
        return iso_utc(local);
    }

    if line == "어제" {
        let y = (local_now - Duration::days(1)).replace_time(Time::MIDNIGHT);
        return iso_utc(y);
    }

    // "YYYY년 M월 D일"
    let nums: Vec<i64> = line
        .split(|c: char| !c.is_ascii_digit())
        .filter_map(|s| s.parse().ok())
        .collect();
    if nums.len() == 3 {
        let month = u8::try_from(nums[1])
            .ok()
            .and_then(|m| time::Month::try_from(m).ok());
        let day = u8::try_from(nums[2]).ok();
        if let (Some(month), Some(day)) = (month, day) {
            if let Ok(date) = time::Date::from_calendar_date(nums[0] as i32, month, day) {
                return iso_utc(date.with_time(Time::MIDNIGHT).assume_offset(local_offset()));
            }
        }
    }

    String::new()
}

/// Current time (used by the live bridge).
pub fn now() -> OffsetDateTime {
    OffsetDateTime::now_utc()
}
